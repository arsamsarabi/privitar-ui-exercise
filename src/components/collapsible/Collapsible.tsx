import React, { FC, ReactElement, useState } from "react";

import {
  StyledCollapsible,
  StyledCollapsibleHeader,
  StyledCollapsibleContent,
} from "./StyledCollapsible";

interface ICollapsibleProps {
  name: string;
  age: number;
  nationality: string;
  privacyRisk: number;
}

export const Collapsible: FC<ICollapsibleProps> = ({
  name,
  age,
  nationality,
  privacyRisk,
}): ReactElement => {
  const [open, setOpen] = useState<boolean>(false);
  return (
    <StyledCollapsible>
      <StyledCollapsibleHeader>
        <p>{name}</p>
        <button onClick={() => setOpen(!open)}>{open ? "-" : "+"}</button>
      </StyledCollapsibleHeader>
      <StyledCollapsibleContent open={open}>
        <p>
          Age: <span>{age}</span>
        </p>
        <div />
        <p>
          Nationality: <span>{nationality}</span>
        </p>
        <p>
          Privacy Risk: <span>{privacyRisk}%</span>
        </p>
      </StyledCollapsibleContent>
    </StyledCollapsible>
  );
};
