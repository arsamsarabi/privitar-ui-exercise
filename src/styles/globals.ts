import { createGlobalStyle, css } from "styled-components";
import { resetCss } from "./reset";

interface GlobalStyleProps {
  theme: ThemeType;
}

export const GlobalStyles = createGlobalStyle<GlobalStyleProps>(({ theme }) => {
  return css`
    ${resetCss}
    body {
      font-family: ${theme.typography.family.primary};
      font-size: 18px;
      font-weight: ${theme.typography.weight.normal};
      color: ${theme.palette.common.grey};
    }
    #root {
      width: 100vw;
      height: 100vh;
      overflow: hidden;
    }
  `;
});
