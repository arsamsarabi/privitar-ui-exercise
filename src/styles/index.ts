export { GlobalStyles } from "./globals";
export { theme } from "./theme";
export { devices } from "./mq";
