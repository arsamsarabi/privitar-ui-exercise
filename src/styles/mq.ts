export const devices = {
  mobile: "(min-width: 375px)",
  tablet: "(min-width: 768px)",
  laptop: "(min-width: 1024px)",
};
