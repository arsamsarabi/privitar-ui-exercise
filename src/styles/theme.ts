import { lighten, darken } from "polished";

const PRIMARY_COLOUR = "#312683";
const SECONDARY_COLOUR = "#af1a63";
const SUCCESS_COLOUR = "#27ae60";
const DANGER_COLOUR = "#e74c3c";

export const theme: ThemeType = {
  typography: {
    family: {
      primary: '"Nunito", Arial, sans-serif',
      secondary: "Sen",
    },
    weight: {
      light: 300,
      normal: 400,
      bold: 700,
    },
  },
  palette: {
    primary: {
      light: lighten(0.25, PRIMARY_COLOUR),
      main: PRIMARY_COLOUR,
      dark: darken(0.25, PRIMARY_COLOUR),
    },
    secondary: {
      light: lighten(0.25, SECONDARY_COLOUR),
      main: SECONDARY_COLOUR,
      dark: darken(0.25, SECONDARY_COLOUR),
    },
    success: {
      light: lighten(0.25, SUCCESS_COLOUR),
      main: SUCCESS_COLOUR,
      dark: darken(0.25, SUCCESS_COLOUR),
    },
    danger: {
      light: lighten(0.25, DANGER_COLOUR),
      main: DANGER_COLOUR,
      dark: darken(0.25, DANGER_COLOUR),
    },
    common: {
      grey: "#575757",
      white: "#ecf0f1",
    },
  },
  config: {
    headerHeight: "48px",
    pagePadding: {
      laptop: {
        v: "24px",
        h: "40px",
      },
    },
    headerPadding: {
      laptop: {
        v: "12px",
        h: "40px",
      },
    },
  },
};
