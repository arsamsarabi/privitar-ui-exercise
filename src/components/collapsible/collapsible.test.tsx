import React from "react";
import { shallow, ShallowWrapper } from "enzyme";

import { Collapsible, ICollapsibleProps } from "./Collapsible";
import {
  StyledCollapsible,
  StyledCollapsibleHeader,
  StyledCollapsibleContent,
} from "./StyledCollapsible";
import { theme } from "../../styles";

describe("Collapsible", () => {
  let wrapper: ShallowWrapper;
  const MOCK_PROPS: ICollapsibleProps = {
    id: 1,
    name: "Eileen Sideways",
    age: 18,
    nationality: "United States",
    privacyRisk: 50,
    expanded: false,
    setExpanded: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(<Collapsible {...MOCK_PROPS} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should render as expected", () => {
    expect(wrapper.find("#person-name").text()).toBe(MOCK_PROPS.name);
    expect(wrapper.find("#collapse-icon").text()).toBe("+");
  });

  it("Should expand/collapse when header is clicked", () => {
    wrapper.find("#collapse-header").simulate("click");
    expect(MOCK_PROPS.setExpanded).toHaveBeenCalledWith(MOCK_PROPS.id);
    setTimeout(() => {
      expect(wrapper.find("#collapse-icon").text()).toBe("-");
      expect(wrapper.find("#person-age").text()).toBe(MOCK_PROPS.age);
      expect(wrapper.find("#person-nationality").text()).toBe(
        MOCK_PROPS.nationality
      );
      expect(wrapper.find("#person-risk-percentage").text()).toBe(
        MOCK_PROPS.privacyRisk
      );
      wrapper.find("#collapse-header").simulate("click");
      expect(MOCK_PROPS.setExpanded).toHaveBeenCalledWith(false);
    }, 100);
  });

  it("Should display '-' for age if age is undefined and same for privacyRisk", () => {
    const wrapper2 = shallow(
      <Collapsible {...MOCK_PROPS} age={undefined} privacyRisk={undefined} />
    );
    wrapper2.find("#collapse-header").simulate("click");
    setTimeout(() => {
      expect(wrapper.find("#person-age").text()).toBe("-");
      expect(wrapper.find("#person-risk-percentage").text()).toBe("-");
    }, 100);
  });
});

describe("StyledCollapsible", () => {
  it("Should render as expected", () => {
    const wrapper = shallow(<StyledCollapsible theme={theme} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("StyledCollapsibleHeader", () => {
  it("Should render as expected", () => {
    const wrapper = shallow(<StyledCollapsibleHeader theme={theme} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("StyledCollapsibleContent", () => {
  it("Should render as expected", () => {
    const wrapper = shallow(<StyledCollapsibleContent theme={theme} />);
    expect(wrapper).toMatchSnapshot();
  });
});
