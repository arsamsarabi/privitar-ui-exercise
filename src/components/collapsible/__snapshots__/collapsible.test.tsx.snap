// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyledCollapsible Should render as expected 1`] = `
.c0 {
  background: #ffffff;
  border: 2px solid #575757;
  padding: 8px 16px;
}

.c0:not(:last-of-type) {
  margin-bottom: 16px;
}

@media (min-width:1024px) {
  .c0:not(:last-of-type) {
    margin-bottom: 24px;
  }
}

<div
  className="c0"
/>
`;

exports[`StyledCollapsibleContent Should render as expected 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-top: 12px;
}

.c0 > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > div p {
  min-width: 50%;
  font-size: 0.95rem;
}

.c0 > div span {
  color: #312683;
  font-weight: 700;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c0 > div:not(:last-of-type) {
  margin-bottom: 8px;
}

@media (min-width:768px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
  }

  .c0 > div {
    width: 50%;
  }

  .c0 > div:first-child {
    margin-right: 50%;
  }

  .c0 > div p {
    font-size: 1.25rem;
  }

  .c0 > div span {
    font-size: 1.35rem;
  }
}

<div
  className="c0"
/>
`;

exports[`StyledCollapsibleHeader Should render as expected 1`] = `
.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 p {
  color: #312683;
  font-weight: 700;
}

.c0 span {
  font-family: 'Carter One',cursive;
  font-size: 1.5rem;
  color: #575757;
}

@media (min-width:768px) {
  .c0 p {
    font-size: 1.25rem;
  }

  .c0 span {
    line-height: 1;
    font-size: 2.5rem;
  }
}

<button
  className="c0"
/>
`;
